File: roml-highs/examples/bess_mpc.rs
//! Direct-Rust BESS rolling MPC benchmark (MPY-05 comparison arm).
//!
//! Same mathematics as the Python `bess_mpc` example, built with the native
//! ROML modeling API and solved through a persistent `Highs` session with
//! batched parameter updates. Prints per-gate wall times and objectives as
//! JSON lines plus a summary: the direct-Rust counterpart for wrapper
//! overhead measurement.
//!
//! Run with: `cargo run -p roml-highs --example bess_mpc --release --
//! --steps 1000 --seed 20260907`

use std::time::Instant;

use roml::prelude::*;
use roml::{ConstraintBounds, ConstraintSpec, LinExpr, SolveOptions, ValueExpr};
use roml_highs::Highs;

const N: usize = 24;
const DT: f64 = 0.25;
const ENERGY_CAP: f64 = 4.0;
const POWER: f64 = 2.0;
const EFF: f64 = 0.95;
const TERMINAL: f64 = 30.0;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = std::env::args().collect();
    let mut steps = 1000usize;
    let mut seed = 20260907u64;
    let mut mode = String::from("closed");
    let mut forecasts_path: Option<String> = None;
    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--steps" => {
                steps = args[i + 1].parse()?;
                i += 2;
            }
            "--seed" => {
                seed = args[i + 1].parse()?;
                i += 2;
            }
            "--mode" => {
                mode = args[i + 1].clone();
                i += 2;
            }
            "--forecasts" => {
                forecasts_path = Some(args[i + 1].clone());
                i += 2;
            }
            other => panic!("unknown argument {other}"),
        }
    }
    if mode != "closed" && mode != "matched" {
        panic!("--mode must be closed or matched");
    }

    let mut model = Model::named("rolling-battery");
    let mut prices = Vec::with_capacity(N);
    for t in 0..N {
        prices.push(model.add_parameter(parameter(50.0).named(format!("price_{t}")))?);
    }
    let initial = model.add_parameter(parameter(2.0).named("initial_energy"))?;
    let mut charge = Vec::with_capacity(N);
    let mut discharge = Vec::with_capacity(N);
    let mut energy = Vec::with_capacity(N + 1);
    let mut direction = Vec::with_capacity(N);
    for t in 0..N {
        charge.push(model.add_variable(continuous().bounds(0.0, POWER).named(format!("c{t}")))?);
        discharge.push(
            model.add_variable(continuous().bounds(0.0, POWER).named(format!("d{t}")))?,
        );
        direction.push(model.add_variable(binary().named(format!("b{t}")))?);
    }
    for t in 0..=N {
        energy.push(
            model.add_variable(continuous().bounds(0.0, ENERGY_CAP).named(format!("e{t}")))?,
        );
    }
    // initial_soc tracks the level parameter: bounds re-applied per gate.
    let initial_soc = model.add_constraint(
        ConstraintSpec::new(
            LinExpr::new().term(1.0, energy[0]),
            ConstraintBounds::eq(2.0),
        )
        .named("initial_soc"),
    )?;
    for t in 0..N {
        let row = LinExpr::new()
            .term(1.0, energy[t + 1])
            .term(-1.0, energy[t])
            .term(-DT * EFF, charge[t])
            .term(DT / EFF, discharge[t]);
        model.add_constraint(ConstraintSpec::new(row, ConstraintBounds::eq(0.0)))?;
        let mode_c = LinExpr::new()
            .term(1.0, charge[t])
            .term(-POWER, direction[t]);
        model.add_constraint(ConstraintSpec::new(mode_c, ConstraintBounds::le(0.0)))?;
        let mode_d = LinExpr::new()
            .term(1.0, discharge[t])
            .term(POWER, direction[t]);
        model.add_constraint(ConstraintSpec::new(mode_d, ConstraintBounds::le(POWER)))?;
    }
    // Objective with parameter-dependent coefficients (native auto-update).
    let mut obj = LinExpr::new();
    for t in 0..N {
        let w = ValueExpr::param(prices[t]) * DT;
        obj = obj.term(w.clone(), discharge[t]);
        obj = obj.term(w * -1.0, charge[t]);
    }
    obj = obj.term(TERMINAL, energy[N]);
    model.maximize(obj)?;

    // Seeded synthetic forecasts (xorshift uniform shaking, same family).
    let mut rng = seed.max(1);
    let mut next_rand = || {
        rng ^= rng << 13;
        rng ^= rng >> 7;
        rng ^= rng << 17;
        (rng as f64) / (u64::MAX as f64)
    };
    let horizon = steps + N;
    let realized: Vec<f64> = match &forecasts_path {
        Some(path) => {
            // Shared frozen stream (one value per line or comma-separated).
            let text = std::fs::read_to_string(path)?;
            text.split(|c| c == ',' || c == '\n')
                .filter_map(|piece| {
                    let piece = piece.trim();
                    if piece.is_empty() {
                        None
                    } else {
                        Some(piece.parse::<f64>().expect("forecast value"))
                    }
                })
                .collect()
        }
        None => {
            let mut realized = Vec::with_capacity(horizon);
            for t in 0..horizon {
                let noise: f64 = (0..12).map(|_| next_rand()).sum::<f64>() - 6.0;
                realized.push(50.0 + 60.0 * ((t as f64) / 4.0).sin() + 5.0 * noise / 2.0);
            }
            realized
        }
    };
    assert!(
        realized.len() >= horizon,
        "forecast stream too short for steps"
    );

    let mut highs = Highs::new()?;
    let options = SolveOptions::new().threads(1).output(false);
    let mut level = 2.0f64;
    let mut walls = Vec::with_capacity(steps);
    for k in 0..steps {
        for t in 0..N {
            model.set_parameter(prices[t], realized[k + t])?;
        }
        let gate_level = if mode == "matched" { 2.0 } else { level };
        model.set_parameter(initial, gate_level)?;
        model.set_constraint_bounds(initial_soc, ConstraintBounds::eq(gate_level))?;
        let t0 = Instant::now();
        let solved = highs.solve_with(&mut model, options.clone())?;
        let wall_ms = t0.elapsed().as_secs_f64() * 1000.0;
        assert!(solved.status().is_optimal());
        let ch0 = solved.value(charge[0]).unwrap_or(0.0);
        let dh0 = solved.value(discharge[0]).unwrap_or(0.0);
        level = (level + DT * (EFF * ch0 - dh0 / EFF)).clamp(0.0, ENERGY_CAP);
        walls.push(wall_ms);
        if k < 3 || k + 1 == steps {
            println!(
                "{{\"gate\":{k},\"wall_ms\":{wall_ms:.3},\"objective\":{:?}}}",
                solved.objective_value()
            );
        }
    }
    walls.sort_by(|a, b| a.partial_cmp(b).unwrap());
    let median = walls[walls.len() / 2];
    let p95 = walls[(walls.len() as f64 * 0.95) as usize];
    println!("{{\"steps\":{steps},\"p50_ms\":{median:.3},\"p95_ms\":{p95:.3}}}");
    Ok(())
}

File: python/examples/bess_mpc.py
"""Rolling-battery MPC example (DESIGN section 8, runnable, self-contained).

Build once, then re-solve over 10 causal forecast gates, applying the
first-interval action each gate. All forecasts are synthetic. Run:
`python python/examples/bess_mpc.py`.
"""

import numpy as np

import roml as rm

N, DT = 24, 0.25
ENERGY_CAP, POWER, EFF, TERMINAL_VALUE = 4.0, 2.0, 0.95, 30.0


def build_model():
    m = rm.Model("rolling-battery")
    price = m.params("price", np.full(N, 50.0))
    initial_energy = m.param("initial_energy", 2.0)
    charge = m.vars("charge", N, ub=POWER)
    discharge = m.vars("discharge", N, ub=POWER)
    energy = m.vars("energy", N + 1, ub=ENERGY_CAP)
    direction = m.vars("direction", N, kind="binary")
    m.add(energy[0] == initial_energy, name="initial_soc")
    m.add(
        energy[1:] == energy[:-1] + DT * (EFF * charge - discharge / EFF),
        name="balance",
    )
    m.add(charge <= POWER * direction, name="charge_mode")
    m.add(discharge <= POWER * (1.0 - direction), name="discharge_mode")
    m.maximize(DT * rm.dot(price, discharge - charge) + TERMINAL_VALUE * energy[-1])
    return m, price, initial_energy, charge, discharge, energy, direction


def main(n_gates=10):
    m, price, initial_energy, charge, discharge, energy, direction = build_model()
    rng = np.random.default_rng(20260907)
    horizon = n_gates + N
    t = np.arange(horizon)
    realized = 50.0 + 60.0 * np.sin(t / 4.0) + rng.normal(0.0, 5.0, size=horizon)
    level = 2.0
    total = 0.0
    with rm.Highs(threads=1, time_limit=2.0) as solver:
        for k in range(n_gates):
            forecasts = realized[k : k + N].copy()
            m.update(price=forecasts, initial_energy=level)
            result = solver.solve(m)
            assert result.has_primal, f"gate {k} has no primal result"
            ch = np.array([result.value(charge[i]) for i in range(N)])
            dh = np.array([result.value(discharge[i]) for i in range(N)])
            en = np.array([result.value(energy[i]) for i in range(N + 1)])
            dr = np.array([result.value(direction[i]) for i in range(N)])
            assert all(min(abs(v), abs(v - 1.0)) <= 1e-6 for v in dr)
            assert abs(en[0] - level) <= 1e-6
            for tt in range(N):
                expect = en[tt] + DT * (EFF * ch[tt] - dh[tt] / EFF)
                assert abs(en[tt + 1] - expect) <= 1e-5
                assert ch[tt] * dh[tt] <= 1e-6
            action = float(dh[0] - ch[0])
            level = float(np.clip(level + DT * (EFF * ch[0] - dh[0] / EFF), 0.0, ENERGY_CAP))
            step_value = float(DT * np.dot(forecasts, dh - ch) + TERMINAL_VALUE * en[-1])
            total += step_value
            print(f"gate {k}: action={action:+.3f} level={level:.3f} value={step_value:.2f}")
    # NOTE: this accumulated total prices forecast energy plus per-gate
    # terminal values; it is a computational smoke signal, not economic
    # evidence of policy quality.
    print(f"total over {n_gates} gates: {total:.2f}")


if __name__ == "__main__":
    main()
